tests/e2e/standard/test_series.py

```diff
@@ -56,34 +66,73 @@ def test_log(self, container: MetadataContainer, series_type: str):
             namespace.assign(FloatSeries(values=values, steps=steps, timestamps=timestamps))
 
     @pytest.mark.parametrize("container", AVAILABLE_CONTAINERS, indirect=True)
     def test_string_series_type_assign(self, container: MetadataContainer):
-        with self.run_then_assert(container, "strings") as (namespace, values, steps, timestamps):
+        with self.run_then_assert_default(container, "strings") as (namespace, values, steps, timestamps):
             namespace.assign(StringSeries(values=values, steps=steps, timestamps=timestamps))
 
     @pytest.mark.parametrize("container", AVAILABLE_CONTAINERS, indirect=True)
     def test_file_series_type_assign(self, container: MetadataContainer):
-        with self.run_then_assert(container, "files") as (namespace, values, steps, timestamps):
+        with self.run_then_assert_default(container, "files") as (namespace, values, steps, timestamps):
             namespace.assign(FileSeries(values=values, steps=steps, timestamps=timestamps))
 
+    @pytest.mark.parametrize("offset,limit,total", NON_LAST_FETCH_CASES)
+    @pytest.mark.parametrize("series_type", ["strings", "floats"])
+    @pytest.mark.parametrize("container", AVAILABLE_CONTAINERS, indirect=True)
+    def test_series_non_last_fetch(
+        self, container: MetadataContainer, series_type: str, offset: int, limit: int, total: int
+    ):
+        fetch_kwargs = {"offset": offset, "limit": limit}
+        steps = list(sorted(random.sample(range(1, 100), total)))
+        timestamps = list(sorted([time.mktime(fake.date_time_this_month().utctimetuple()) for _ in range(total)]))
+
+        if series_type == "floats":
+            values = [random.random() for _ in range(total)]
+            series_init = FloatSeries
+        elif series_type == "strings":
+            values = [fake.word() for _ in range(total)]
+            series_init = StringSeries
+        elif series_type == "files":
+            values = [generate_image(size=2**n) for n in range(7, 7 + total)]
+            series_init = FileSeries
+        else:
+            raise ValueError(f"Unknown series type: {series_type}")
+
+        expected_steps = steps[offset : offset + limit]
+        expected_timestamps = timestamps[offset : offset + limit]
+        expected_values = values[offset : offset + limit]
+
+        with self.run_then_assert_with_expected(
+            container,
+            series_type,
+            expected_steps,
+            expected_timestamps,
+            expected_values,
+            last_value=values[-1],
+            fetch_kwargs=fetch_kwargs,
+        ) as (namespace, _, _, _):
+            namespace.assign(series_init(values=values, steps=steps, timestamps=timestamps))
+
     @contextmanager
-    def run_then_assert(self, container: MetadataContainer, series_type: str):
+    def run_then_assert_default(
+        self, container: MetadataContainer, series_type: str, fetch_kwargs: Dict[str, Any] = {}
```

Comment:
Mutable defaults are not a good idea. Perhaps we could pass `None` as default and then set to `{}` inside the function body?

Reply:
good point!
Forgot about that completely